CoreScriptsRoot/Modules/VR/LaserPointer.lua

```diff
@@ -0,0 +1,548 @@
+
+	local parabHitPart, parabHitPoint, parabHitNormal, parabHitMaterial, t = self.parabola:FindPartOnParabola(ignore)
+	return parabHitPart, parabHitPoint, parabHitNormal, t
+end
+
+function LaserPointer:getLaserHit(pos, look, ignore)
+	local ray = Ray.new(pos, look * LASER.MAX_DISTANCE)
+	local laserHitPart, laserHitPoint, laserHitNormal, laserHitMaterial = workspace:FindPartOnRayWithIgnoreList(ray, ignore)
+	local t = (laserHitPoint - pos) / LASER.MAX_DISTANCE
+	return laserHitPart, laserHitPoint, laserHitNormal, t
+end
+
+function LaserPointer:setTeleportMaxRangePerSecond(maxRange)
+	self.teleportMaxRangePerSecond = maxRange
+end
+
+function LaserPointer:calculateLaunchVelocity(gravity, launchAngle)
+	local maxVelocity = TELEPORT.MAX_VALID_DISTANCE
+	local minVelocity = TELEPORT.MIN_VELOCITY
+	local velocityRange = math.max(0, maxVelocity - minVelocity)
+	return ((self.teleportRangeT ^ TELEPORT.RANGE_T_EXP) * velocityRange) + minVelocity
+end
+
+function LaserPointer:recomputePath(startPos, endPos)
+	self.computingPath = true
+
+	self.pathStart = startPos
+	self.pathEnd = endPos
+	coroutine.wrap(function()
+		self.lastPath = PathfindingService:ComputeRawPathAsync(startPos, endPos, TELEPORT.MAX_VALID_DISTANCE)
+		self.computingPath = false
+	end)()
+end
+
+function LaserPointer:checkLastPath()
+	if not self.lastPath then
+		self.pathValid = true
+		return
+	end
+	if self.lastPath.Status ~= Enum.PathStatus.Success then
+		self.pathValid = false
+		return
+	end
+	local occludedPoint = self.lastPath:CheckOcclusionAsync(0)
+	if occludedPoint < 0 then
+		self.pathValid = true
+	end
+end
+
+function LaserPointer:canTeleportTo(cameraPos, part, point, normal)
+	local character = LocalPlayer.Character
+	if not character then
+		return false
+	end
+	local humanoidRootPart = character:FindFirstChild("HumanoidRootPart")
+	if not humanoidRootPart then
+		return false
+	end
+
+	if not part then
+		return false
+	end
+	if normal.Y < 0 then
+		return false
+	end
+
+	local dist = (point - cameraPos).magnitude
+	if dist > TELEPORT.MAX_VALID_DISTANCE then
+		return false
+	end
+
+	local bbPos = point
+	local halfBbSize = TELEPORT.CLEAR_AABB_SIZE * 0.5
+	local minBound = Vector3.new(bbPos.X - halfBbSize.X, bbPos.Y, bbPos.Z - halfBbSize.Z)
+	local maxBound = minBound + TELEPORT.CLEAR_AABB_SIZE
+
+	local theta = math.rad(90) - math.asin(normal.Y)
+	local slopeOffset = math.sin(theta) * math.sqrt(halfBbSize.X^2 + halfBbSize.Z^2)
+	self.collisionTestPart.CFrame = CFrame.new(point + Vector3.new(0, (TELEPORT.CLEAR_AABB_SIZE.Y / 2) + slopeOffset + 0.1, 0))
+	
+	--Not only does workspace:FindPartsInRegion3 only do AABBs, it also doesn't seem to like slopes.
+	--Using an arbitary part here sounds nasty as an implementation detail, but the concept seems sound.
+	--Getting the part out of the workspace entirely would be a good long-term solution.
+	local foundParts = self.collisionTestPart:GetTouchingParts()
```

Comment:
I can try that. I don't know if we really want to call it FindPartsInRegion3 at that point though; maybe we could add a new method in Workspace that takes a CFrame and Size instead. 

Region3s already have a CFrame but there's no Lua way of setting that right now. We might be able to just add a constructor overload that takes a CFrame and a Size to Region3, but we'll also need to make sure you're not passing rotated Region3s to Terrain methods.